src/main/java/com/celements/cleverreach/CleverReachRest.java

```diff
@@ -0,0 +1,305 @@
+package com.celements.cleverreach;
+
+import static com.celements.model.util.References.*;
+import static com.google.common.base.Preconditions.*;
+
+import java.io.IOException;
+import java.util.Map;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation.Builder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.NewCookie;
+import javax.ws.rs.core.Response;
+
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.annotation.InstantiationStrategy;
+import org.xwiki.component.annotation.Requirement;
+import org.xwiki.component.descriptor.ComponentInstantiationStrategy;
+import org.xwiki.context.Execution;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.SpaceReference;
+
+import com.celements.auth.classes.RemoteLoginClass;
+import com.celements.cleverreach.exception.CleverReachRequestFailedException;
+import com.celements.model.access.IModelAccessFacade;
+import com.celements.model.access.exception.DocumentNotExistsException;
+import com.celements.model.classes.ClassDefinition;
+import com.celements.model.context.ModelContext;
+import com.celements.model.object.xwiki.XWikiObjectFetcher;
+import com.google.common.base.Optional;
+import com.sun.syndication.io.impl.Base64;
+import com.xpn.xwiki.objects.BaseObject;
+
+@Component(CleverReachRest.COMPONENT_NAME)
+@InstantiationStrategy(ComponentInstantiationStrategy.SINGLETON)
+public class CleverReachRest implements CleverReachService {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CleverReachRest.class);
+
+  public static final String COMPONENT_NAME = "rest";
+
+  public static final String REST_CONFIG_SPACE_NAME = "Configs";
+  public static final String REST_CONFIG_DOC_NAME = "CleverReachRest";
+  protected static final String DEFAULT_REST_URL = "https://rest.cleverreach.com/";
+  static final String PATH_VERSION = "v3/";
+  static final String PATH_LOGIN = "oauth/token.php";
+  static final String PATH_MAILINGS = "mailings.json/";
+  static final String PATH_WHOAMI = "debug/whoami.json";
+  static final String PATH_TTL = "debug/ttl.json";
+
+  static final String CONTEXT_CONNECTION_KEY = "clever_reach_connection";
+
+  static enum SubmitMethod {
+    GET, POST, PUT, DELETE
+  };
+
+  @Requirement
+  private IModelAccessFacade modelAccess;
+
+  @Requirement
+  private ModelContext modelContext;
+
+  @Requirement(RemoteLoginClass.CLASS_DEF_HINT)
+  private ClassDefinition remoteLoginClass;
+
+  @Requirement
+  private Execution execution;
+
+  Client injectedClient;
```

Comment:
Instead you could create a IRestClientFactoryRole with createClient() method and implement the "default" Component using the ClientBuilder.newClient() this would allow to Mock the factory in tests as usual.